src/PubSubClient.cpp

```diff
@@ -329,25 +329,48 @@ size_t PubSubClient::readPacket(uint8_t* lenLen) {
  */
-void PubSubClient::handlePacket(uint8_t llen, size_t len) {
+void PubSubClient::handlePacket(uint8_t hdrLen, size_t length) {
     uint8_t type = this->buffer[0] & 0xF0;
     DEBUG_PSC_PRINTF("received message of type %u\n", type);
     switch (type) {
         case MQTTPUBLISH:
             if (callback) {
-                uint8_t* payload;
-                uint16_t tl = (this->buffer[llen + 1] << 8) + this->buffer[llen + 2]; /* topic length in bytes */
-                memmove(this->buffer + llen + 2, this->buffer + llen + 3, tl);        /* move topic inside buffer 1 byte to front */
-                this->buffer[llen + 2 + tl] = 0;                                      /* end the topic as a 'C' string with \x00 */
-                char* topic = (char*)this->buffer + llen + 2;
-                // msgId only present for QOS>0
-                if ((this->buffer[0] & 0x06) == MQTTQOS1) {
-                    uint16_t msgId = (this->buffer[llen + 3 + tl] << 8) + this->buffer[llen + 3 + tl + 1];
-                    payload = this->buffer + llen + 3 + tl + 2;
-                    callback(topic, payload, len - llen - 3 - tl - 2);
+                // MQTT Publish packet: See section 3.3 MQTT v3.1 protocol specification:
+                // - Header: 1 byte
+                // - Remaining header length: hdrLen bytes, multibyte field (1 .. MQTT_MAX_HEADER_SIZE - 1)
+                // - Topic length: 2 bytes (starts at buffer[hdrLen + 1])
+                // - Topic: topicLen bytes (starts at buffer[hdrLen + 3])
+                // - Packet Identifier (msgId): 0 bytes for QoS 0, 2 bytes for QoS 1 and 2 (starts at buffer[hdrLen + 3 + topicLen])
+                // - Payload (for QoS = 0): length - (hdrLen + 3 + topicLen) bytes (starts at buffer[hdrLen + 3 + topicLen])
+                // - Payload (for QoS > 0): length - (hdrLen + 5 + topicLen) bytes (starts at buffer[hdrLen + 5 + topicLen])
+                // To get a null reminated 'C' topic string we move the topic 1 byte to the front (overwriting the LSB of the topic lenght)
+                uint16_t topicLen = (this->buffer[hdrLen + 1] << 8) + this->buffer[hdrLen + 2];  // topic length in bytes
+                char* topic = (char*)(this->buffer + hdrLen + 3 - 1);  // set the topic in the LSB of the topic lenght, as we move it there
+                uint16_t payloadOffset = hdrLen + 3 + topicLen;        // payload starts after header and topic (if there is no packet identifier)
+                size_t payloadLen = length - payloadOffset;            // this might change by 2 if we have a QoS 1 or 2 message
+                uint8_t* payload = this->buffer + payloadOffset;
+
+                if (length < payloadOffset) {  // do not move outside the max bufferSize
+                    ERROR_PSC_PRINTF_P("handlePacket(): Suspicious topicLen (%u) points outside of received buffer length (%zu)\n", topicLen, length);
+                    return;
+                }
```

Comment:
Maybe it's better to stop the connection instead just return ...
```
                if (length < payloadOffset) {  // do not move outside the max bufferSize
                    _state = MQTT_DISCONNECTED;
                    _client->stop();
                    ERROR_PSC_PRINTF_P("handlePacket(): Suspicious topicLen (%u) points outside of received buffer length (%zu)\n", topicLen, length);
                    return;
```

Reply:
Or you could let this function return a bool signalling success and act on the return value of this function call.

Reply:
Good idea, I'll do that. But still I am not sure if this is the correct reaction on wrong data in the message stream ...

The current implementation just crashes here (I haven't seen this, but it seems to happen). 
@uschindler What's your opinion on that?

Reply:
Ok, update is done. What do you think?

Reply:
It is the same reaction to the broker as the broker will do when you send something unexpected.
For example start sending a publish message telling you will be sending N bytes.
If you send anything more you will be disconnected.
On sending less, you will be disconnected due to timeout or incorrect next header.